### PR TITLE
autocommit added to connection object

### PR DESCRIPTION
**Pull request by yafimvo**
## Changes
`autocommit` option added to the SQL connection object

closes: #90

### CODE REVIEW

File: src/sql/run.py [hunk 1]
```diff
@@ -391,6 +391,10 @@ def run(conn, sql, config, user_namespace):
                 result = FakeResultProxy(cur, headers)
             else:
                 txt = sqlalchemy.sql.text(statement)
+
+                if config.autocommit:
+                    conn.session.execution_options(isolation_level="AUTOCOMMIT")
+
```
Comment (edublancas) [inline src/sql/run.py:397]: is this supported by all databases? my guess is that it's not and it'll break in some cases

Comment (yafimvo) [reply src/sql/run.py:397]: According to sqlalchemy [this is how we set the autocommit option](https://docs.sqlalchemy.org/en/20/orm/session_transaction.html#setting-transaction-isolation-levels-dbapi-autocommit). 

I think the question is if someone is setting `sql.autocommit = True` - is it the same as setting the autocommit option

Comment (edublancas) [reply src/sql/run.py:397]: True, I think we need to dig deeper into this, possibly in sqlalchemy's docs

Comment (edublancas) [reply src/sql/run.py:397]: for example, the error doesn't happen in SQLite, I can create a table without issues

Comment (edublancas) [reply src/sql/run.py:397]: let's wait for https://github.com/ploomber/jupysql/issues/116 so we have a better idea on how this might affect other databases

also, I looked at the code. based on what I saw, I guess that the original author didn't use `execution_options(isolation_level="AUTOCOMMIT")` because it's [not compatible with all dbs](https://docs.sqlalchemy.org/en/20/core/connections.html#setting-isolation-level-or-dbapi-autocommit-for-a-connection), instead it's executing a commit command after running SQL.

Comment (tonykploomber) [reply src/sql/run.py:397]: Test case added in https://github.com/ploomber/jupysql/pull/117 👍

File: src/sql/run.py [hunk 2]
```diff
@@ -400,8 +400,8 @@ def run(conn, sql, config, user_namespace):
                         conn.session.execution_options(isolation_level="AUTOCOMMIT")
                     except Exception as e:
                         print(
-                            "The database driver do not support \
-                            such execution option", e
+                            "The database driver do not support "
```
Comment (yafimvo) [inline src/sql/run.py:403]: don't support -> doesn't support
(same in test)

**Review by yafimvo — COMMENTED** [review]
@edublancas what did we decide on this "autocommit" execution?

**Review by idomic — COMMENTED** [review]
Small comment

File: src/tests/integration/test_duckDB.py [hunk 3]
```diff
@@ -0,0 +1,5 @@
+def test_auto_commit_mode_on(ip_with_duckDB, capsys):
```
[the remark this answers is not in the document]
Comment (tonykploomber) [reply src/tests/integration/test_duckDB.py:1]: Reasonable, will cover it

File: src/sql/run.py (hunk 2 above)
Comment (edublancas) [reply src/sql/run.py:403]: > @edublancas what did we decide on this "autocommit" execution?

see the latest paragraph [here](https://github.com/ploomber/jupysql/issues/90#issuecomment-1420628820)

tl;dr; use the native autocommit if possible, otherwise fallback to the current implementation (issuing  a COMMIT command manually)

also @idomic FYI, this is what I proposed as a solution

File: src/tests/integration/test_duckDB.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tonykploomber) [reply src/tests/integration/test_duckDB.py:1]: Add in [f749b14](https://github.com/ploomber/jupysql/pull/91/commits/f749b1477ea948a899ec6ac2b85ff6fa53560c84)

The case is about `%config SqlMagic.autocommit=False`, which should disable the autocommit and not show any errors

**Comment by edublancas** [comment]
is this ready for review? @tonykploomber

**Comment by tonykploomber** [comment]
> is this ready for review? @tonykploomber

Yes

File: src/sql/run.py [hunk 4]
```diff
@@ -390,8 +395,17 @@ def run(conn, sql, config, user_namespace):
                 result = FakeResultProxy(cur, headers)
             else:
                 txt = sqlalchemy.sql.text(statement)
+                if config.autocommit:
+                    try:
+                        conn.session.execution_options(isolation_level="AUTOCOMMIT")
+                    except Exception as e:
+                        print(
+                            "The database driver doesn't support "
+                            "such AUTOCOMMIT execution option", "\nException: ", e
```
Comment (edublancas) [inline src/sql/run.py:404]: let's add something at the end "running manual COMMIT command" so users know what we're doing

if my assumption is right, then looks like we're missing some unit tests

File: src/sql/run.py [hunk 5]
```diff
@@ -358,11 +358,15 @@ def fetchmany(size):
...
     """Issues a commit, if appropriate for current config and dialect"""
 
-    _should_commit = config.autocommit and all(
-        dialect not in str(conn.dialect) for dialect in _COMMIT_BLACKLIST_DIALECTS
+    _should_commit = (
+        config.autocommit
+        and all(
+            dialect not in str(conn.dialect) for dialect in _COMMIT_BLACKLIST_DIALECTS
+        )
+        and not manual_commit
```
Comment (edublancas) [inline src/sql/run.py:369]: shouldn't this be `and manual_commit`? looks like that controls whether we should or not run the commit command

File: src/tests/integration/test_duckDB.py [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+    ip_with_duckDB.run_cell("%sql CREATE TABLE weather (city VARCHAR,);")
+    out, _ = capsys.readouterr()
+    assert "The database driver doesn't support such AUTOCOMMIT execution option" in out
+
+
+def test_auto_commit_mode_off(ip_with_duckDB, capsys):
+    ip_with_duckDB.run_cell("%config SqlMagic.autocommit=False")
+    ip_with_duckDB.run_cell("%sql CREATE TABLE weather (city VARCHAR,);")
+    out, _ = capsys.readouterr()
+    assert (
```
Comment (edublancas) [inline src/tests/integration/test_duckDB.py:12]: better to check for the output of the confirmation message instead of looking that the error message doesn't exist

File: src/tests/integration/test_postgreSQL.py [hunk 7]
```diff
@@ -7,8 +6,6 @@ def test_meta_cmd_display(ip_with_postgreSQL):
     assert ("public", "taxi", "table", "ploomber_app") in out.result
 
 
-# Known issue, addressing in https://github.com/ploomber/jupysql/issues/90
-@pytest.mark.xfail(reason="known autocommit mode issue")
 def test_auto_commit_mode_on(ip_with_postgreSQL, capsys):
     ip_with_postgreSQL.run_cell("%config SqlMagic.autocommit=True")
     ip_with_postgreSQL.run_cell("%sql CREATE DATABASE new_db")
```
Comment (edublancas) [inline src/tests/integration/test_postgreSQL.py:11]: run_cell won't raise any exceptions even if the code fails; we must ensure that the property that holds the exc is empty to guarantee that there were no. Also, we should check that the output of the `create database` command prints some confirmation message to ensure that the command executed successfully

File: src/sql/run.py (hunk 4 above)
Comment (edublancas) [reply src/sql/run.py:404]: also, it should be clear what part of the error message is from our side and which from the driver's side. I'm seeing this:

```
The database driver doesn't support such AUTOCOMMIT execution option 
Exception:  'duckdb.DuckDBPyConnection' object has no attribute 'set_isolation_level'
(duckdb.NotImplementedException) Not implemented Error: CREATE DATABASE not supported in DuckDB yet
[SQL: CREATE DATABASE new_db]
(Background on this error at: https://sqlalche.me/e/14/tw8g)
```

but would be better to have a format like this:

```
The database driver doesn't support such AUTOCOMMIT execution option 

Message from the database driver:
    Exception:  'duckdb.DuckDBPyConnection' object has no attribute 'set_isolation_level' (duckdb.NotImplementedException) Not implemented Error: CREATE DATABASE not supported in DuckDB yet
     [SQL: CREATE DATABASE new_db]
    (Background on this error at: https://sqlalche.me/e/14/tw8g)
```

File: src/sql/run.py (hunk 5 above)
Comment (edublancas) [reply src/sql/run.py:369]: also, let's add a message so the user knows when the commit failed:

so change this:

```python
except sqlalchemy.exc.OperationalError:
    pass # not all engines can commit
```

to something like this:
```python
except sqlalchemy.exc.OperationalError:
    # not all engines can commit
   print("The database does not support the COMMIT command")
```

File: CHANGELOG.md [hunk 8]
```diff
@@ -2,6 +2,8 @@
 
 ## 0.5.7dev
 
+* [Fix] Addresses enable AUTOCOMMIT config issue in PostgreSQL ([#90](https://github.com/ploomber/jupysql/issues/90))
```
Comment (edublancas) [inline CHANGELOG.md:5]:
```suggestion
* [Fix] Fixes AUTOCOMMIT config issue in PostgreSQL #90
```

(no need to add the link, this is added automatically by our deployment script)

**Review by edublancas — CHANGES_REQUESTED** [review]

**Comment by tonykploomber** [comment]
> @yafimvo any updates on this one? It's been pending for a while, let's try to wrap it up.

Let me fix those suggestions now

File: src/sql/run.py (hunk 4 above)
Comment (tonykploomber) [reply src/sql/run.py:404]: Updated the message to 
```
The database driver doesn't support such AUTOCOMMIT execution option
Perhaps you can try running a manual COMMIT command

Message from the database driver: 
	Exception: 'duckdb.DuckDBPyConnection' object has no attribute 'set_isolation_level' 

(duckdb.NotImplementedException) Not implemented Error: CREATE DATABASE not supported in DuckDB yet
[SQL: CREATE DATABASE new_db2]
(Background on this error at: https://sqlalche.me/e/14/tw8g)
```

(The last section starting from `duckdb.NotImplementedException` is out of control of this code, where we can't control the indent here, but I guess now it's better)
Screenshot 2023-03-01 at 2 04 37 PM

File: src/tests/integration/test_postgreSQL.py (hunk 7 above)
Comment (tonykploomber) [reply src/tests/integration/test_postgreSQL.py]: Fix in [70021af](https://github.com/ploomber/jupysql/pull/91/commits/70021afdb86e2caa3829a5de4e825469b07db206)

Changes: 
1. Check exec has any error
2. the output of the `create database` command doesn't throw any message, we run the manual listing databases command `\l` and assert if new databases gets created here

**Comment by tonykploomber** [comment]
> @tonykploomber updates here?

Ready for review again 
@yafimvo

**Review by edublancas — APPROVED** [review]
